polar-core/src/resource_block.rs

```diff
@@ -280,21 +282,61 @@ impl ResourceBlocks {
         resource: Term,
         declarations: Declarations,
         shorthand_rules: Vec<ShorthandRule>,
-    ) {
-        self.declarations.insert(resource.clone(), declarations);
-        self.shorthand_rules
-            .insert(resource.clone(), shorthand_rules);
+    ) -> Vec<ValidationError> {
+        let mut errors = vec![];
+
+        // Merge existing declarations if we are reopening a resource block, otherwise add new
+        if let Some(existing) = self.declarations.get_mut(&resource) {
+            for (key, new) in &declarations {
+                let existing_declaration = existing.get(key);
+                // allow
+                // 1. overwriting declarations of the **same** type
+                // 2. inserting declarations where none previously existed
+                //
+                // deny overwriting of declarations with mismatched types
+                match (new, existing_declaration) {
+                    // new or same-type declarations are accepted
+                    (Declaration::Relation(_), Some(Declaration::Relation(_)))
```

Comment:
Doesn't this miss the case where I've declared `relations = { a: A };` in one block and `relations = { a: B };` in another?

Reply:
This is handled in dd9dd3f4 but it'd still be good to add a test case for this

Reply:
Added a test case for relations of differing types in 160e72aa91ae3bbb7fb06be1d51491190775166c